crates/bevy_ecs/src/entity/allocator.rs

```diff
@@ -0,0 +1,966 @@
+pub(crate) struct AllocUniqueEntityRowIterator(core::ops::Range<u32>);
+
+impl Iterator for AllocUniqueEntityRowIterator {
+    type Item = EntityRow;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0
+            .next()
+            // SAFETY: This came from an *exclusive* range. It can never be max.
+            .map(|idx| unsafe { EntityRow::new(NonMaxU32::new_unchecked(idx)) })
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.0.size_hint()
+    }
+}
+
+impl ExactSizeIterator for AllocUniqueEntityRowIterator {}
+impl core::iter::FusedIterator for AllocUniqueEntityRowIterator {}
+
+/// An [`Iterator`] returning a sequence of [`Entity`] values from an [`Allocator`].
+///
+/// **NOTE:** Dropping will leak the remaining entities!
+pub struct AllocEntitiesIterator<'a> {
+    new: AllocUniqueEntityRowIterator,
+    reused: FreeBufferIterator<'a>,
+}
+
+impl<'a> Iterator for AllocEntitiesIterator<'a> {
+    type Item = Entity;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.reused
+            .next()
+            .or_else(|| self.new.next().map(Entity::from_raw))
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let len = self.reused.len() + self.new.len();
+        (len, Some(len))
+    }
+}
+
+impl<'a> ExactSizeIterator for AllocEntitiesIterator<'a> {}
+impl<'a> core::iter::FusedIterator for AllocEntitiesIterator<'a> {}
+
+// SAFETY: Newly reserved entity values are unique.
+unsafe impl EntitySetIterator for AllocEntitiesIterator<'_> {}
+
+impl Drop for AllocEntitiesIterator<'_> {
+    fn drop(&mut self) {
+        let leaking = self.len();
+        if leaking > 0 {
+            warn!(
+                "{} entities being leaked via unfinished `AllocEntitiesIterator`",
+                leaking
+            );
+        }
+    }
+}
+
+/// This is a stripped down version of [`Allocator`] that operates on fewer assumptions.
+/// As a result, using this will be slower than [`Allocator`] but this offers additional freedoms.
+#[derive(Clone)]
+pub struct RemoteAllocator {
+    shared: Arc<SharedAllocator>,
+}
+
+impl RemoteAllocator {
+    /// Creates a new [`RemoteAllocator`] with the provided [`Allocator`] source.
+    /// If the source is ever destroyed, [`Self::alloc`] will yield garbage values.
+    /// Be sure to use [`Self::is_closed`] to determine if it is safe to use these entities.
+    pub fn new(source: &Allocator) -> Self {
+        Self {
+            shared: source.shared.clone(),
+        }
+    }
+
+    /// Allocates an entity remotely.
+    #[inline]
+    pub fn alloc(&self) -> Entity {
+        self.shared.remote_alloc()
+    }
+
+    /// Returns whether or not this [`RemoteAllocator`] is still connected to its source [`Allocator`].
+    /// Note that this could close immediately after the function returns false, so be careful.
+    pub fn is_closed(&self) -> bool {
+        self.shared.is_closed()
+    }
```

Comment:
So is the correct pattern here to eagerly allocate and then check `is_closed` to see if there's a chance you got garbage?